src/Router.js

```diff
@@ -85,62 +70,119 @@ export default class Router {
       throw new Error('Router#registerRoutes must be passed an array of Routes')
     }
     routes.map(route => {
-      this.__routes.push(new Route(route))
+      route = new Route(route);
+      if (this.onRouteComplete) {
+        const handlerLength = route.handlers.length-1;
+        const lastHandler = route.handlers[handlerLength];
+        const routingEnd = (ctx, next) => {
+          lastHandler(ctx);
+          const duration = fns.getNow() - this.__startTime;
+          const fromPath = this.__fromPath || 'PAGE LOAD';
+          this.onRouteComplete({
+            fromPath,
+            toPath: this.__currentCanonicalPath,
+            duration,
+          })
+        };
+        route.handlers = route.handlers.slice(0, handlerLength).concat(routingEnd);
+      }
+      if (this.onRouteStart) {
+        route.handlers.unshift(this.onRouteStart);
+      }
+      this.__routes.push(route);
     })
   }
 
   /**
+   * @param {String} path
+   */
+  registerCatchallPath(path) {
+    this.__catchallPath = path;
+  }
+
+  /**
    * @param {String} canonicalPath
    */
   go(canonicalPath) {
-    this.__dispatch(canonicalPath, true)
+    this.__dispatch(canonicalPath, false)
   }
 
   /**
    * @param {String} canonicalPath
    */
   replace(canonicalPath) {
-    this.__dispatch(canonicalPath, false)
+    this.__dispatch(canonicalPath, true)
   }
 
   reset() {
-    this.__routes = []
+    this.__catchallPath = null;
+    this.__routes = [];
+    WindowEnv.removeEventListener('popstate', this.__onpopstate)
```

Comment:
I wonder if `reset` should reset the route start and route end functions.  what do you think?